README
Apricots-0.2.6

Overview

        Apricots is a game I've been writing in my spare time.
        Despite the name, it has nothing to do with apricots.
        It's a game where you fly a little plane around the
        screen and shoot things and drop bombs on enemy targets,
        and it's meant to be quick and fun.

History

        The original version of the game dates back to around
        1991. The very first version of the game was written on
        an Amiga, using the AMOS language, and the game was
        called 'Planegame'. This was a very simple affair with
        simplistic landscapes, no guns or shrapnel, and very
        limited computer AI. It was also dreadfully slow. The
        AMOS compiler improved things somewhat at a later date. A
        couple of years later I obtained the BLITZ language for
        the Amiga, and ported 'Planegame' carefully over to that,
        resulting in a huge speedup. Impressed by the capabilites
        of BLITZ, in 1995 I proceeded to rewrite the game
        entirely from scratch, even using completely new graphics
        and sound. The result was 'Planegame2', which, for
        various reasons also got dubbed 'Apricots'. This new
        version of the game added anti-aircraft guns, shrapnel,
        trees, towers, water, better airbases and a huge
        improvement on graphics, sound, and intelligent computer
        AI. Slightly later, in 1996, the Draks were added as a
        bit of amusement. The result was a highly playable game,
        which I distributed to a few of my friends. A few years
        later and I now use a PC, which can run Linux. Having
        learnt C++ by this time, at the beginning of 2002 I felt
        ready enough to port 'Apricots' cross platform from BLITZ
        on the Amiga to C++ on a Linux PC. What you see here is
        work in progress on that port. The game has been tweaked
        slightly while being ported over, so is not identical to
        the Amiga version.

Authors

        Project Leader: Mark Snellgrove (Original concept and game
                        coding)
        Co-author: Mark Harman (First Windows port and SDL conversion)
        Additional contributor: Judebert (Configuration file)
                                           
Status
        The game is still under development. Apricots-0.2.6 is a 
        beta release using the SDL library. Despite this, it
        is highly playable. Please mail apricots@fishies.org.uk
        any bug reports, upgrades or comments.

C++ Compilers

        Apricots is known to successfully build under the following
        OS/Compiler configurations
        
        Linux - gcc (reported by Mark Snellgrove)
        OpenBSD - gcc (reported by Johan Luff)
        Windows - Borland (reported by Mark Harman)
        Windows - Visual C++ 6.0 (reported by Mark Snellgrove)
        Windows/Cygwin - gcc (reported by Judebert)
        
        It may build successfuly on other platforms, but this has
        not been tested. Let me know!
                
Requirements

            SDL > 1.2.0, download from www.libsdl.org
            OpenAL, download from www.openal.org

Compilation options

        -DAP_PATH="datadir" Set directory where apricots looks for
                            its data files. Defaults to current directory
                            if unset.
                        
Instructions

        CONTROLS                      PLAYER 1             PLAYER 2
        -----------------------------------------------------------
        ROTATE LEFT                   CURSOR LEFT          Z
        ROTATE RIGHT                  CURSOR RIGHT         C
        ACCELERATE / STEALTH          CURSOR UP            S
        DROP BOMB                     CURSOR DOWN          X
        FIRE SHOT                     RETURN               LEFT CTRL
        
        Gameplay is fairly common sense to work out.
        Stay alive, and shooting enemy targets gains you points,
        whereas being shot down or shooting the wrong things lose
        you points. You have infinite planes, and are competing
        against the other computer planes (or other player!).
        First to a certain score wins once they land at base.
        Simple!

Play Options

        At present there is no option screen. Options can be changed by
        editing the configuration file apricots.cfg (on UNIX systems this
        will probably require root access, user configuration files will
        be added soon in a future release). The number and type of planes
        can be varied, along with the airbases, as can 1/2 player game
        be chosen. The number of anti aircraft guns and scenery
        can be selected. Draks can be turned on or off, and a
        different condition for winning the game can be chosen.

License

        Apricots falls under the GPL. Please read the file COPYING included
        in this distrubution. The sourcecode of apricots-0.2.6 will either
        be included in the distrubution, or can otherwise be downloaded
        from the apricots webpage (see below)

Webpage

        For updates, bugfixes, binaries, sourcecode and more info please
        visit the apricots webpage at
        http://www.fishies.org.uk/apricots.html


NEWS
* 2023-05-11 moggers87  0.2.8

- Make OpenAL a hard requirement
- Migrate from freealut to alure
- Fix clamping issue with newer versions of SDL2
- A trio of code fixes from Dmitry Marakasov (thanks!)

* 2021-02-12 moggers87 0.2.7

- Ported to SDL2
- Cleaned up code meet modern C++ standards
- Changelog is now generate from `git log`
- Game now scales properly on larger screens
- Configuration is now searched for in multiple places:
  - ~/.config/apricots/
  - System config directory (defaults to /usr/local/etc)
  - Finally, defaults are stored with game data

* 2019-08-03 moggers87

- Applied Fedora patches
- Added to git repo
- Regenerated autoconf files without KDE macros

* 2003-08-13 Mark Snellgrove 0.2.6

- Version 0.2.6 released

* 2003-08-08 Mark Snellgrove

- Sanity check on configuration options now applied
- Configuration from apricots.cfg extended to all options at runtime
- Portability Bugfixes applied to getConfig()
- Mouse pointer now hidden during game
- Major code cleanup in sampleio class. Looks tidier, but BUG1 still present
- BORLAND specific options in apricots.cpp found to be redundant - Removed.
- Game speed now properly regulated by TICK_INTERVAL parameter (Fixes BUG2)

* 2003-08-01 Judebert

- Limited Configuration read at runtime from apricots.cfg
- Windows port of OpenAL audio routines in sampleio class
- Cygwin compiler flag now checked on main

* 2003-07-30 Mark Snellgrove 0.2.5

- Version 0.2.5 released
- File configure.in now checks for OpenAL (thanks to Jens Thiele)
- Now using ANSI-C++ standard includes
- Header cstdlib included in SDLfont.h
- Functions max() and min() replaced by limit() in function draw_dither()
- Windows specific options _WIN32 changed to Borland specific BORLAND in apricots.cpp
- Several loop variables renamed to allow for evil non-ANSI scoping rules in MSVC++
- Hardware surfaces now used (SDL_HWSURFACE) replacing software surfaces
- New struct smoketype declared in apricots.h
- Smoke list now changed to be of type linkedlist <smoketype>
- Bugfix so smoke now obeys GAME_SPEED parameter correctly
- Bugfix so bomb y-velocity now obeys GAME_SPEED parameter correctly
- AI modified to account for change in bomb trajectory due to above
- Bugfix now inputs font file with ios::binary flag set
- Options in init_gamedata() function are now fully commented
- Global parameter GAME_SPEED reduced to 0.2

* 2003-07-26 Mark Snellgrove 0.2.4

- Version 0.2.4 released
- Audio via OpenAL reactivated (via conditional compiler option)

* 2003-07-24 Mark Snellgrove 0.2.3

- Version 0.2.3 released
- SDLfont palettes and masks pre-rendered for speedup

* 2003-07-17 Mark Snellgrove

- SDLfont class for rendering fonts in SDL implemented (slow)

* 2003-07-13 Mark Snellgrove 0.2.2

- Version 0.2.2 released
- Existence of apricots.shapes file checked
- Airbase constructor included (required by Borland compiler)
- Several minor bugs fixed in shape class

* 2003-07-09 Mark Snellgrove 0.2.1

- Version 0.2.1 released
- Airbase constructors modified
- Bugfix in shape constructor
- Palette bluescale modified for dithering (SDL uses 5 bit blue palette)
- Player 2 controls altered and reactivated
- Player 2 playfield fixed
- Compilation warnings from implicit casts fixed
- Compilation warnings in null building definition fixed

* 2003-07-06 Mark Snellgrove 0.2.0

- Version 0.2.0 released (SDL version)
- Fixed up some Linux/Windows portability issues
- rand() now used as random number generator
- Font routines temporarily disabled (portability problem)

* 2003-06-21 Mark Harman

- Windows port using the SDL graphics library
- OpenAL Audio temporarily disabled (portability problem)
- Controls for player 2 disabled.

* 2002-09-13 Mark Snellgrove 0.1.0

- Version 0.1.0 released (SVGALIB/OpenAL version)

* 2002-03-26 Mark Snellgrove

- SVGALIB/OpenAL version of apricots in playable form.


apricots/init.cpp
// Apricots init routine
// Author: M.D.Snellgrove
// Date: 26/3/2002
// History:

// Changes by M Harman for Windows version, June 2003:
//   Changes for graphics and font related stuff.
//   Need to open "apricots.shapes" file in ios::binary mode.

// Changes by M Snellgrove 8/7/2003
//   Palette bluescale redefined (in SDL this appears to be 5 bit, not 6 bit)

// Changes by M Snellgrove 13/7/2003
//   Check for existence of apricots.shapes file

// Changes by Judebert 1/8/2003
//   Some configuration now read from apricots.cfg

// Changes by M Snellgrove 8/8/2003
//   Cursor hidden
//   All configuration now read from apricots.cfg, with error checking

#include "apricots.h"

// Display setup (double buffered with two playfields)

void setup_display(gamedata &g) {
  SDL_Window *window =
      SDL_CreateWindow("Apricots", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, SCREEN_WIDTH, SCREEN_HEIGHT,
                       SDL_WINDOW_ALLOW_HIGHDPI | SDL_WINDOW_MAXIMIZED | SDL_WINDOW_RESIZABLE);

  g.renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_SOFTWARE);
  if (g.renderer == NULL) {
    fprintf(stderr, "Couldn't create renderer: %s\n", SDL_GetError());
    exit(1);
  }

  if (SDL_RenderSetLogicalSize(g.renderer, SCREEN_WIDTH, SCREEN_HEIGHT) != 0) {
    fprintf(stderr, "Couldn't set logical resizing: %s\n", SDL_GetError());
    exit(1);
  }

  g.virtualscreen = SDL_CreateRGBSurface(SDL_SWSURFACE, SCREEN_WIDTH, SCREEN_HEIGHT, 8, 0, 0, 0, 0);
  if (g.virtualscreen == NULL) {
    fprintf(stderr, "Couldn't set 640x480x8 virtual video mode: %s\n", SDL_GetError());
    exit(1);
  }

  g.gamescreen = SDL_CreateRGBSurface(SDL_SWSURFACE, GAME_WIDTH, GAME_HEIGHT, 8, 0, 0, 0, 0);
  if (g.gamescreen == NULL) {
    fprintf(stderr, "Couldn't set game video mode: %s\n", SDL_GetError());
    exit(1);
  }
}

// Load font and set font colours

void load_font(SDL_Surface *screen, SDLfont &whitefont, SDLfont &greenfont) {

  char filename[255];
  strcpy(filename, AP_PATH);
  strcat(filename, "alt-8x16.psf");
  whitefont.loadpsf(filename, 8, 16);
  whitefont.colour(screen, 1, 0);
  greenfont = whitefont;
  greenfont.colour(screen, 13, 0);
}

// Load shapes and set palette

void load_shapes(gamedata &g, shape images[]) {

  char filename[255];
  strcpy(filename, AP_PATH);
  strcat(filename, "apricots.shapes");
  ifstream fin(filename, ios::binary);
  if (fin.fail()) {
    fprintf(stderr, "Could not open file: %s\n", filename);
    exit(-1);
  }

  {
    char dummy[14];
    fin.read(dummy, 14);
    int dummyb;
    fin >> dummyb;
    int red, green, blue;
    for (int c1 = 0; c1 < 256; c1++) {
      SDL_Color col = {0, 0, 0, 0};
      SDL_SetPaletteColors(g.virtualscreen->format->palette, &col, c1, 1);
    }
    for (int c2 = 0; c2 < 16; c2++) {
      fin >> red >> green >> blue;
      Uint8 new_red = red * 4;
      Uint8 new_green = green * 4;
      Uint8 new_blue = blue * 4;
      SDL_Color col = {new_red, new_green, new_blue, 0};
      SDL_SetPaletteColors(g.virtualscreen->format->palette, &col, c2, 1);
    }
    for (int c3 = 0; c3 < 25; c3++) {
      Uint8 new_green = ((2 * c3) % 64) * 4;
      Uint8 new_blue = ((15 + 2 * c3) % 64) * 4;
      SDL_Color col = {0, new_green, new_blue, 0};
      SDL_SetPaletteColors(g.virtualscreen->format->palette, &col, c3 + 16, 1);
    }
    for (int c4 = 0; c4 < 256; c4++) {
      Uint8 rgb[3];
      SDL_GetRGB(c4, g.virtualscreen->format, &rgb[0], &rgb[1], &rgb[2]);
      SDL_Color col = {rgb[0], rgb[1], rgb[2], 0};
      SDL_SetPaletteColors(g.gamescreen->format->palette, &col, c4, 1);
    }
    fin.read(dummy, 1);
  }
  for (int i = 1; i <= 34; i++)
    images[i].read(fin);
  for (int j = 69; j <= 318; j++)
    images[j].read(fin);

  for (int c = 0; c < 256; c++) {
    Uint8 rgb[3];
    SDL_GetRGB(c, g.virtualscreen->format, &rgb[0], &rgb[1], &rgb[2]);
    SDL_Color col = {rgb[0], rgb[1], rgb[2], 0};
    for (int i = 0; i <= 318; i++) {
      if (images[i].getSurface() != NULL)
        SDL_SetPaletteColors(images[i].getSurface()->format->palette, &col, c, 1);
    }
  }

  fin.close();
}

// Sound initialization

void init_sound(sampleio &sound) {

  char filenames[14][255];
  for (int i = 0; i < 14; i++) {
    strcpy(filenames[i], AP_PATH);
  }
  strcat(filenames[0], "engine.wav");
  strcat(filenames[1], "jet.wav");
  strcat(filenames[2], "explode.wav");
  strcat(filenames[3], "groundhit.wav");
  strcat(filenames[4], "fuelexplode.wav");
  strcat(filenames[5], "shot.wav");
  strcat(filenames[6], "gunshot.wav");
  strcat(filenames[7], "bomb.wav");
  strcat(filenames[8], "splash.wav");
  strcat(filenames[9], "laser.wav");
  strcat(filenames[10], "stall.wav");
  strcat(filenames[11], "gunshot2.wav");
  strcat(filenames[12], "afterburner.wav");
  strcat(filenames[13], "finish.wav");

  sound.init(14, filenames, 2, 6);
}

// Initialize the game constants

void init_gameconstants(gamedata &g) {

  // Initialize sin/cos arrays and afterburner co-ordinates
  for (int i = 1; i <= 16; i++) {
    g.xmove[i] = -sin((i - 1) * PI / 8.0);
    g.ymove[i] = -cos((i - 1) * PI / 8.0);
    g.xboost[i] = 6 - int(9 * g.xmove[i]);
    g.yboost[i] = 6 - int(10 * g.ymove[i]);
  }
  // Corrections
  g.ymove[5] = 0.0;
  g.ymove[13] = 0.0;
  g.xmove[1] = 0.0;
  g.xmove[9] = 0.0;
  g.xboost[1] = 5;
  g.xboost[5] = 17;
  g.yboost[5] = 7;
  g.xboost[13] = -5;
  g.yboost[13] = 7;
  // Initialize flight data
  g.accel[1] = -0.18;
  g.accel[2] = -0.12;
  g.accel[3] = -0.04;
  g.accel[4] = -0.005;
  g.accel[5] = 0.0;
  g.accel[6] = 0.02;
  g.accel[7] = 0.05;
  g.accel[8] = 0.15;
  g.accel[9] = 0.2;
  g.accel[10] = 0.15;
  g.accel[11] = 0.05;
  g.accel[12] = 0.02;
  g.accel[13] = 0.0;
  g.accel[14] = -0.005;
  g.accel[15] = -0.04;
  g.accel[16] = -0.12;
  // Initialize the bomb images
  g.bombimage[1] = 115;
  g.bombimage[2] = 122;
  g.bombimage[3] = 122;
  g.bombimage[4] = 122;
  g.bombimage[5] = 121;
  g.bombimage[6] = 120;
  g.bombimage[7] = 120;
  g.bombimage[8] = 120;
  g.bombimage[9] = 119;
  g.bombimage[10] = 118;
  g.bombimage[11] = 118;
  g.bombimage[12] = 118;
  g.bombimage[13] = 117;
  g.bombimage[14] = 116;
  g.bombimage[15] = 116;
  g.bombimage[16] = 116;
}

//--JAM: Gets a line from the config string with format:
//  NAME:VALUE\n
// and returns it as a string.  If there is no line for the
// name, returns the default.

string getConfig(string config, string name, string defval) {
  // Pull out just the name line
  size_t ndx = config.find(name);

  if (ndx == string::npos) {
    return defval;
  }
  ndx = config.find(":", ndx);

  // Advance past spaces
  while (config.at(++ndx) == ' ') {
  };
  if (ndx == string::npos) {
    return defval;
  }

  return config.substr(ndx, config.find("\n", ndx) - ndx);
}

// Similar to above, but this returns integers, and includes error checking.

int getConfig(string config, string name, int defval, int min, int max) {
  // Pull out just the name line
  size_t ndx = config.find(name);

  if (ndx == string::npos) {
    return defval;
  }
  ndx = config.find(":", ndx);

  // Advance past spaces
  while (config.at(++ndx) == ' ') {
  };
  if (ndx == string::npos) {
    return defval;
  }

  int value = strtol(config.substr(ndx, config.find("\n", ndx) - ndx).c_str(), 0, 10);

  // Bounds checking
  if ((value < min) || (value > max)) {
    cerr << "Entry " << name.c_str() << " out of bounds in apricots.cfg" << endl;
    cerr << name.c_str() << " must take values between " << min << " and " << max << endl;
    exit(EXIT_FAILURE);
  }

  return value;
}

// Initialize the game parameters
// Edit the values here to set game options

string find_config_file() {
  string filename("apricots.cfg");

  std::filesystem::path user_path("/");
  const char *xdg_config_home = std::getenv("XDG_CONFIG_HOME");
  if (xdg_config_home == NULL) {
    const char *home_dir = std::getenv("HOME");
    if (home_dir == NULL) {
      fprintf(stderr, "Neither XDG_CONFIG_HOME nor HOME were set, unable to locate per user configuration\n");
    } else {
      user_path /= home_dir;
      user_path /= ".config";
    }
  } else {
    user_path /= xdg_config_home;
  }
  user_path /= "apricots";
  user_path /= filename;

  std::filesystem::path sysconfig_path(SYSCONFIG_PATH);
  sysconfig_path /= filename;

  std::filesystem::path data_path(AP_PATH);
  data_path /= filename;

  if (std::filesystem::exists(user_path)) {
    return user_path;
  } else if (std::filesystem::exists(sysconfig_path)) {
    return sysconfig_path;
  } else {
    return data_path;
  }
}

void init_gamedata(gamedata &g) {

  string filename = find_config_file();
  ifstream config_stream(filename);
  //--JAM: Read from config file
  string config;
  if (!config_stream.fail()) {
    // Read config file line by line
    char line[256];
    while (!config_stream.eof()) {
      config_stream.getline(line, 255);
      if (line[0] != '#') {
        config.append(line);
        config += "\n";
      }
    }
  } else {
    // Config file not found
    config = "\n";
  }
  config_stream.close();

  // Number of planes (1-6)
  g.planes = getConfig(config, "NUM_PLANES", 2, 1, 6);

  // Number of players (1 or 2)
  g.players = getConfig(config, "NUM_HUMANS", 1, 1, 2);
  // Error check
  if (g.players > g.planes) {
    cerr << "Invalid configuration in apricots.cfg" << endl;
    cerr << "Number of human players cannot exceed number of planes" << endl;
    exit(EXIT_FAILURE);
  }

  // Mission
  // 0/1 means winner is first to reach targetscore and reach airbase
  // 1 makes score reduce to targetscore-200 upon dying (so must land)
  // 2 means winner is first to destroy all enemy airbases and reach airbase
  g.mission = getConfig(config, "MISSION", 0, 0, 2);

  // Targetscore for missions 0/1
  g.targetscore = getConfig(config, "TARGET_SCORE", 1400, 100, 5000);

  // Planetypes: 1=Spitfire, 2=Jet, 3=Stealth Bomber
  g.planeinfo[1].planetype = getConfig(config, "PLANE1", 1, 1, 3);
  g.planeinfo[2].planetype = getConfig(config, "PLANE2", 1, 1, 3);
  g.planeinfo[3].planetype = getConfig(config, "PLANE3", 1, 1, 3);
  g.planeinfo[4].planetype = getConfig(config, "PLANE4", 1, 1, 3);
  g.planeinfo[5].planetype = getConfig(config, "PLANE5", 1, 1, 3);
  g.planeinfo[6].planetype = getConfig(config, "PLANE6", 1, 1, 3);

  // Basetype: See create_airbases in setup.cpp
  g.planeinfo[1].basetype = getConfig(config, "BASE1", 1, 1, 7);
  g.planeinfo[2].basetype = getConfig(config, "BASE2", 1, 1, 7);
  g.planeinfo[3].basetype = getConfig(config, "BASE3", 1, 1, 7);
  g.planeinfo[4].basetype = getConfig(config, "BASE4", 1, 1, 7);
  g.planeinfo[5].basetype = getConfig(config, "BASE5", 1, 1, 7);
  g.planeinfo[6].basetype = getConfig(config, "BASE6", 1, 1, 7);

  // Control: 1=Player 1, 2=Player 2, 0=AI
  g.planeinfo[1].control = getConfig(config, "CONTROL1", 1, 0, 2);
  g.planeinfo[2].control = getConfig(config, "CONTROL2", 0, 0, 2);
  g.planeinfo[3].control = getConfig(config, "CONTROL3", 0, 0, 2);
  g.planeinfo[4].control = getConfig(config, "CONTROL4", 0, 0, 2);
  g.planeinfo[5].control = getConfig(config, "CONTROL5", 0, 0, 2);
  g.planeinfo[6].control = getConfig(config, "CONTROL6", 0, 0, 2);
  // Error check
  int count[3];
  count[0] = 0;
  count[1] = 0;
  count[2] = 0;
  for (int i = 1; i <= g.planes; i++) {
    count[g.planeinfo[i].control]++;
  }
  if (count[1] != 1) {
    cerr << "Invalid configuration in apricots.cfg" << endl;
    cerr << "Invalid control selection for player 1" << endl;
    exit(EXIT_FAILURE);
  }
  if (count[2] != (g.players - 1)) {
    cerr << "Invalid configuration in apricots.cfg" << endl;
    cerr << "Invalid control selection for player 2" << endl;
    exit(EXIT_FAILURE);
  }
  for (int j = g.planes + 1; j <= 6; j++) {
    if (g.planeinfo[j].control > 0) {
      cerr << "Invalid configuration in apricots.cfg" << endl;
      cerr << "Human controls specified for non-playing plane" << endl;
      exit(EXIT_FAILURE);
    }
  }
  // Number of towerblocks
  g.towers = getConfig(config, "NUM_TOWERS", 5, 0, 30);

  // Number of neutral anti-aircraft guns
  g.guns = getConfig(config, "NUM_GUNS", 5, 0, 20);

  // Number of other buildings
  g.buildings = getConfig(config, "NUM_BUILDINGS", 20, 0, 50);

  // Number of trees (max)
  g.trees = getConfig(config, "NUM_TREES", 50, 0, 100);

  // Draks: 0=Never, 1=5% probability, 2=Always
  string drakval = getConfig(config, "DRAK", "sometimes");
  if (!drakval.compare("always")) {
    g.drakoption = 2;
  } else if (!drakval.compare("sometimes")) {
    g.drakoption = 1;
  } else {
    g.drakoption = 0;
  }
}

// Main Initialization routine

void init_data(gamedata &g) {

  init_gameconstants(g);

  init_gamedata(g);

  // Set Random seed
  srand(time(0));

  /* Initialize defaults, Video and Audio */
  if ((SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO) == -1)) {
    fprintf(stderr, "Could not initialize SDL: %s.\n", SDL_GetError());
    exit(-1);
  }

  setup_display(g);

  // Hide cursor
  SDL_ShowCursor(0);

  load_shapes(g, g.images);

  load_font(g.virtualscreen, g.whitefont, g.greenfont);

  init_sound(g.sound);
}
